From c66be452da92259193e05f59f85755765afcb092 Mon Sep 17 00:00:00 2001
From: "Vishnu Swaroop Kumar Sarma, Duddu" <duddu.swaroop@intel.com>
Date: Wed, 24 Jul 2019 00:59:19 +0530
Subject: [PATCH] UGW_SW-41261 sso gpio blink bit/src update fix

For the software update of the sso gpios, the blink_r bit should be 0.
Also the mask passed is uninitialised in that case which is fixed now.
---
 drivers/gpio/gpio-intel-sso.c | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git a/drivers/gpio/gpio-intel-sso.c b/drivers/gpio/gpio-intel-sso.c
index 5b26140297c0..b1b31ca4cbd2 100644
--- a/drivers/gpio/gpio-intel-sso.c
+++ b/drivers/gpio/gpio-intel-sso.c
@@ -179,10 +179,11 @@ static int sso_gpio_freq_set(struct sso_gpio_priv *priv)
 
 	freq_idx = sso_gpio_get_freq_idx(priv->freq);
 	val = freq_idx % FPID_FREQ_RANK_MAX;
+	off = 0;
+	mask = 0;
 
 	if (!priv->freq) {
 		us = US_SW;
-		off = US;
 	} else if (freq_idx < FPID_FREQ_RANK_MAX) {
 		mask = SSO_CON1_FPID_MASK;
 		off = SSO_CON1_FPID;
@@ -193,9 +194,15 @@ static int sso_gpio_freq_set(struct sso_gpio_priv *priv)
 		us = US_GPTC;
 	}
 
-	/* Update FSC/GPT Divider and US for LED  */
+	/* Update FSC/GPT Divider/BLINK_R and US for LED  */
 	sso_gpio_write_mask(priv->mmap, SSO_CON1, US, US_MASK, us);
-	sso_gpio_write_mask(priv->mmap, SSO_CON1, off, mask, val);
+	/* Update BLINK_R and update src only for non software update */
+	if(priv->freq) {
+		sso_gpio_update_bit(priv->mmap, SSO_CON0, BLINK_R, 1);
+		sso_gpio_write_mask(priv->mmap, SSO_CON1, off, mask, val);
+	} else {
+		sso_gpio_update_bit(priv->mmap, SSO_CON0, BLINK_R, 0);
+	}
 
 	return 0;
 }
@@ -281,9 +288,6 @@ static int sso_gpio_hw_init(struct sso_gpio_priv *priv)
 	if (sso_gpio_update_bit(priv->mmap, SSO_CON0, RZFL, priv->edge))
 		return -ENOTSUPP;
 
-	if (sso_gpio_update_bit(priv->mmap, SSO_CON0, BLINK_R, 1))
-		return -ENOTSUPP;
-
 	/* Set GPIO update rate */
 	sso_gpio_freq_set(priv);
 
